### ISSUE TITLE
Question about generation results.

### ISSUE THREAD

**Issue by sangHa0411** [huggingface/text-generation-inference#127]
I compared generation sentences between `huggingface/text-generation inference` method and `generate function` which AutoModelForCausalLM model has. 

I put same seed, top_k, top_p, typical_p and do_sample arguments, but these two methods generate different outputs though they got same input sentence.

I want to know if it is normal or I tested wrong.

Thank you!

**Comment by OlivierDehaene** [comment]
- What model were you using?
- Were you using a GPU?
- What launch parameters did you use?

**Comment by sangHa0411** [comment]
Thanks for your answer!

* I used "facebook/galactica-6.7b" model.
* I used NVIDIA A100 GPU.
* This is my launch parameter
  ```bash
  model=facebook/galactica-6.7b
  num_shard=1
  max_input_length=2048
  max_total_tokens=2048
  volume=/home/kyungmin.lee/sangha/data
  gpus='"device=0"'
  
  docker run -it --gpus $gpus \
      -p 8000:80 \
      -v $volume:/data ghcr.io/huggingface/text-generation-inference:latest \
      --model-id $model \
      --num-shard $num_shard \
      --max-input-length $max_input_length \
      --max-total-tokens $max_total_tokens
  ```

* I use random seed 0. These are generation codes I used.
    * model.gernerate()
        ```python
        torch.manual_seed(0)
        torch.cuda.manual_seed(0)
        torch.cuda.manual_seed_all(0)
        torch.backends.cudnn.deterministic = True
        torch.backends.cudnn.benchmark = False
        
        output_sentences = []
        
        for sen in tqdm(input_sentences) :
            sen = sen.strip()
            
            with torch.no_grad() :
        
                example_input = tokenizer(sen, return_tensors='pt')
                example_input = example_input.to(device)
                
                example_input_ids = example_input["input_ids"]
        
                generation_results = model.greedy_search(
                    input_ids=example_input_ids, 
                    max_length=output_seqeunce_length,
                    min_length=output_seqeunce_length,
                )
            
            generation_sentence = tokenizer.batch_decode(generation_results, skip_special_tokens=True)[0]
            output_sentences.append(generation_sentence)
        ```
    * text-generation inference
      ```python
      for sen in tqdm(input_sentences) :
          input_sen = sen.strip()
          
          input_ids = tokenizer(input_sen, return_tensors='pt')['input_ids'][0]
          
          generation_sentence = client.generate(
              input_sen, 
              max_new_tokens=output_seqeunce_length-len(input_ids),
              seed=0,       
              return_full_text=True
          ).generated_text
      ```

**Comment by jshin49** [comment]
This issue seems relevant to what I'm going through. In my case, even when I give `do_sample=False` as a parameter, the generation results seem to differ each time.

**Comment by sangHa0411** [comment]
This issue happened because GPU device.
I used a100 gpu which support bf16, then bf16 was used automatically when I used huggingface/text-generation-inference.
So the results from my server (a100 gpu) was different from HuggingFace outputs (FP32).
When I used fp32 with this repository, results from `huggingface/text-generation inference method `and `generate function which AutoModelForCausalLM model has` are same.

I hope this finding helpful to you. 

@jshin49
